### ISSUE TITLE
Failing to open test_index dir when using tantivy/master IndexReader

### ISSUE THREAD

**Issue by petr-tik** [tantivy-search/tantivy-py#14]
@karlicoss told me he tried to install tantivy-py on his machine last night and got this error, so I decided to investigate tonight and found a repro. 

```bash
petr_tik@merluza:~/Coding/tantivy-py$ RUST_BACKTRACE=1 python3 -m pytest --capture=no
================================================= test session starts =================================================
platform linux -- Python 3.6.8, pytest-5.0.1, py-1.7.0, pluggy-0.12.0
rootdir: /home/petr_tik/Coding/tantivy-py
collected 15 items

tests/tantivy_test.py .......thread '<unnamed>' panicked at 'attempt to subtract with overflow', /home/petr_tik/.cargo/git/checkouts/tantivy-cd4bdf03c5df22c3/4c39417/src/directory/footer.rs:73:55
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:59
             at src/libstd/panicking.rs:197
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:211
   4: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:474
   5: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:381
   6: rust_begin_unwind
             at src/libstd/panicking.rs:308
   7: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
   8: core::panicking::panic
             at src/libcore/panicking.rs:49
   9: tantivy::directory::footer::Footer::from_bytes
             at /home/petr_tik/.cargo/git/checkouts/tantivy-cd4bdf03c5df22c3/4c39417/src/directory/footer.rs:73
  10: tantivy::directory::footer::Footer::extract_footer
             at /home/petr_tik/.cargo/git/checkouts/tantivy-cd4bdf03c5df22c3/4c39417/src/directory/footer.rs:85
  11: <tantivy::directory::managed_directory::ManagedDirectory as tantivy::directory::directory::Directory>::open_read
             at /home/petr_tik/.cargo/git/checkouts/tantivy-cd4bdf03c5df22c3/4c39417/src/directory/managed_directory.rs:250
  12: tantivy::core::segment::Segment::open_read
             at /home/petr_tik/.cargo/git/checkouts/tantivy-cd4bdf03c5df22c3/4c39417/src/core/segment.rs:80
  13: tantivy::core::segment_reader::SegmentReader::open
             at /home/petr_tik/.cargo/git/checkouts/tantivy-cd4bdf03c5df22c3/4c39417/src/core/segment_reader.rs:149
  14: core::ops::function::FnMut::call_mut
             at /rustc/50a0defd5a93523067ef239936cc2e0755220904/src/libcore/ops/function.rs:148
  15: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
             at /rustc/50a0defd5a93523067ef239936cc2e0755220904/src/libcore/ops/function.rs:279
  16: core::option::Option<T>::map
             at /rustc/50a0defd5a93523067ef239936cc2e0755220904/src/libcore/option.rs:416
  17: <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::next
             at /rustc/50a0defd5a93523067ef239936cc2e0755220904/src/libcore/iter/adapters/mod.rs:575
  18: <<core::result::Result<V,E> as core::iter::traits::collect::FromIterator<core::result::Result<A,E>>>::from_iter::Adapter<Iter,E> as core::iter::traits::iterator::Iterator>::next
             at /rustc/50a0defd5a93523067ef239936cc2e0755220904/src/libcore/result.rs:1250
  19: <&mut I as core::iter::traits::iterator::Iterator>::next
             at /rustc/50a0defd5a93523067ef239936cc2e0755220904/src/libcore/iter/traits/iterator.rs:2607
  20: <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T,I>>::from_iter
             at /rustc/50a0defd5a93523067ef239936cc2e0755220904/src/liballoc/vec.rs:1819
  21: <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
             at /rustc/50a0defd5a93523067ef239936cc2e0755220904/src/liballoc/vec.rs:1731
  22: <core::result::Result<V,E> as core::iter::traits::collect::FromIterator<core::result::Result<A,E>>>::from_iter
             at /rustc/50a0defd5a93523067ef239936cc2e0755220904/src/libcore/result.rs:1267
  23: core::iter::traits::iterator::Iterator::collect
             at /rustc/50a0defd5a93523067ef239936cc2e0755220904/src/libcore/iter/traits/iterator.rs:1465
  24: tantivy::reader::InnerIndexReader::reload
             at /home/petr_tik/.cargo/git/checkouts/tantivy-cd4bdf03c5df22c3/4c39417/src/reader/mod.rs:128
  25: tantivy::reader::IndexReaderBuilder::try_into
             at /home/petr_tik/.cargo/git/checkouts/tantivy-cd4bdf03c5df22c3/4c39417/src/reader/mod.rs:71
  26: tantivy::core::index::Index::reader
             at /home/petr_tik/.cargo/git/checkouts/tantivy-cd4bdf03c5df22c3/4c39417/src/core/index.rs:201
  27: tantivy::index::Index::new
             at src/index.rs:185
  28: tantivy::index::__init15304180475000063864::__init15304180475000063864::__wrap::{{closure}}
             at src/index.rs:155
  29: tantivy::index::__init15304180475000063864::__init15304180475000063864::__wrap
             at src/index.rs:155
  30: <unknown>
  31: _PyObject_FastCallKeywords
  32: <unknown>
  33: _PyEval_EvalFrameDefault
  34: _PyFunction_FastCallDict
  35: <unknown>
  36: PyObject_Call
  37: _PyEval_EvalFrameDefault
  38: <unknown>
  39: <unknown>
  40: PyObject_Call
  41: _PyEval_EvalFrameDefault
  42: <unknown>
  43: <unknown>
  44: <unknown>
  45: _PyEval_EvalFrameDefault
  46: <unknown>
  47: <unknown>
  48: <unknown>
  49: _PyEval_EvalFrameDefault
  50: <unknown>
  51: <unknown>
  52: _PyEval_EvalFrameDefault
  53: <unknown>
  54: _PyFunction_FastCallDict
  55: <unknown>
  56: <unknown>
  57: _PyObject_FastCallKeywords
  58: <unknown>
  59: _PyEval_EvalFrameDefault
  60: <unknown>
  61: <unknown>
  62: _PyEval_EvalFrameDefault
  63: <unknown>
  64: <unknown>
  65: PyObject_Call
  66: _PyEval_EvalFrameDefault
  67: <unknown>
  68: <unknown>
  69: <unknown>
  70: _PyEval_EvalFrameDefault
  71: <unknown>
  72: <unknown>
  73: <unknown>
  74: _PyEval_EvalFrameDefault
  75: <unknown>
  76: <unknown>
  77: _PyEval_EvalFrameDefault
  78: <unknown>
  79: _PyFunction_FastCallDict
  80: <unknown>
  81: <unknown>
  82: PyObject_Call
  83: _PyEval_EvalFrameDefault
  84: <unknown>
  85: <unknown>
  86: <unknown>
  87: _PyEval_EvalFrameDefault
  88: <unknown>
  89: <unknown>
  90: <unknown>
  91: _PyEval_EvalFrameDefault
  92: <unknown>
  93: <unknown>
  94: PyObject_Call
  95: _PyEval_EvalFrameDefault
  96: <unknown>
  97: <unknown>
  98: <unknown>
  99: _PyEval_EvalFrameDefault
fatal runtime error: failed to initiate panic, error 5
Fatal Python error: Aborted

Current thread 0x00007fea645c6740 (most recent call first):
  File "/home/petr_tik/Coding/tantivy-py/tests/tantivy_test.py", line 148 in test_opens_from_dir
```

line 148 points to:

```python
        index = Index(schema(), PATH_TO_INDEX, reuse=True)
```

which relies on IndexReader opening the directory. Our test_index was built with tantivy 0.10, before the introduction of directory footer. Since then tantivy-py switched to tracking the master branch of tantivy. 

This shows the value of having integration tests with a serialised index. 

I [suggested keeping the same version numbers with core tantivy](https://github.com/tantivy-search/tantivy-py/issues/6) for ease of administration, but it must have been skipped since.

**Comment by fulmicoton** [comment]
> I suggested keeping the same version numbers with core tantivy for ease of administration, but it must have been skipped since.

I don't understand what this means here.

There has been a discussion on whether tantivy-py master should point to tantivy master branhc or tantivy revision. @poljar prefers master to avoid last minute integration. There are pros and cons on both sides, let's respect his choice.

The issue you experienced is not due to the publication of 0.10.2. The introduction of the footer is not cargo published yet. It breaks compatibility so it can only happen for 0.11.

> This shows the value of having integration tests with a serialised index.

I would argue the exact opposite and recommend removing the serialized index.

One last note. 
The error sucks, but the good news is that in the future, you will get a proper error message telling you that the version the index was built with is incompatible with your tantivy version, and the actual name of the version used.

**Comment by poljar** [comment]
I just noticed this yesterday as well and tried to mitigate the problem using `catch_unwind()`. 

> The error sucks, but the good news is that in the future, you will get a proper error message telling you that the version the index was built with is incompatible with your tantivy version, and the actual name of the version used.

Does that mean that it'll be turned from a panic into an error? If so, that's great news since catching the panic doesn't seem to be possible because `UnsafeCell` is used somewhere further down in the call stack.

**Comment by fulmicoton** [comment]
@fdb-hiroshima  could you confirm my claim?

**Comment by fdb-hiroshima** [comment]
As things are, there is an Enum that allow to detect if the index is incompatible, but it does not propagate an error, nor does it log anything actually

**Comment by fulmicoton** [comment]
@fdb-hiroshima thanks.

If I understand correctly, we just do not check fro the compatibility today, but there is no fundamental reasons that would prevent us from doing it right?

I opened tantivy#662 to address that before the 0.11 release.

**Comment by fdb-hiroshima** [comment]
That's right

**Comment by petr-tik** [comment]
My original comment was rash, as it was getting quite late that day. 

I appreciate and accept the decision taken previously. I thought that having a failing integration test with a directory index was a good early failure to show to the users. If we decide to remove it, it would be great to have a way to fail tests and not let users deploy a shared library that will silently corrupt or error in prod. 

My previous suggestion was to keep the same number version in rust tantivy and tantivy-py. If we are going to track master, I expect we will need to test that every commit in master works with the rest of tantivy-py and uncovers failures otherwise. AFAIK, we might be able to set up GitHub actions to run tantivy-py CI on every commit to the master branch of tantivy

The alternative is to track tantivy upgrades on crates and go through a potentially big merge/upgrade in tantivy-py. 

I hope this clarifies my earlier comments. 

Whatever we decide, I would like to optimise the user experience. This includes python programmers with little experience and time to debug broken shared libraries and read stacktraces.

**Comment by poljar** [comment]
Your previous suggestion was taken into consideration, at least how I understood it. The version numbers will be kept the same between tanitvy and tantivy-py. What you might have missed, and I probably failed to mention before, is that we cannot depend on a released version of tantivy for now.

That is one of the reasons why there hasn't been a release of tantivy-py yet and why users are required to use the repo version of it. I wanted tantivy-py to be set up to track the master branch of tantivy to uncover issues like this one before a release happens. 

It should minimize the chance of tantivy-py being stuck with a sub-optimal released version of tantivy and possibly having to wait for another release.

I agree that we should optimize the user experience, and we should work towards a release to make that happen. Since there was no release users are probably aware that this is a work in progress and issues might arise.

**Comment by petr-tik** [comment]
> What you might have missed, and I probably failed to mention before, is that we cannot depend on a released version of tantivy for now.

You are right - I missed this. That makes sense now. 

> That is one of the reasons why there hasn't been a release of tantivy-py yet and why users are required to use the repo version of it. I wanted tantivy-py to be set up to track the master branch of tantivy to uncover issues like this one before a release happens.

I am glad we did with the index. @fulmicoton prefers we remove the serialised index from the tests, so would be good to keep the ability to catch those errors. 

> I agree that we should optimize the user experience, and we should work towards a release to make that happen. Since there was no release users are probably aware that this is a work in progress and issues might arise.

I found your API v easy to use, so I employed it in my demo/presentation that I did about tantivy in Rust London. Several people asked me about it and I even recommended it to a friend of mine. I should have asked you about its status. 

I would only suggest making the WIP nature of tantivy-py more explicit in the Readme until our release. 

Please let me know, how I can contribute to the release of tantivy-py!

**Comment by poljar** [comment]
So I added a milestone and the issues that we probably want to resolve before doing a release. Of course a tantivy relase that exposes the missing types that we are using in the bindings should also happen before that (unless we really want do to a release that depends on a pinned github commit).

The API simplification is probably one of the more interesting issues on the list and comments are always welcome there.

**Comment by fulmicoton** [comment]
> Of course a tantivy relase that exposes the missing types that we are using in the bindings should also happen before

@poljar Absolutely. What is the current plan : are we going to ship binaries? If you have a targetted schedule, let me know so I can land tantivy 0.11 before that.

**Comment by fulmicoton** [comment]
> What is the current plan : are we going to ship binaries?

Ooops sorry, you are discussing that on #4

**Comment by poljar** [comment]
Binaries as well as a source distribution are planned, I didn't yet test out building binaries that can be published on pypi but as far as I checked out the documentation it should be possible.

I don't have a targeted schedule, whenever tantivy 0.11 releases is fine by me.

**Comment by poljar** [comment]
Fixed by 2d6ecff430a6a28f15c6b39e713caff3b80fb2ed.
